scene/gui/color_picker.cpp

```diff
@@ -353,27 +677,77 @@ void ColorPicker::_value_changed(double) {
 		return;
 	}
 
-	if (hsv_mode_enabled) {
-		h = scroll[0]->get_value() / 360.0;
-		s = scroll[1]->get_value() / 100.0;
-		v = scroll[2]->get_value() / 100.0;
-		if (picker_type == SHAPE_OKHSL_CIRCLE) {
-			color.set_ok_hsl(h, s, v, Math::round(scroll[3]->get_value() / 255.0));
-		} else {
-			color.set_hsv(h, s, v, Math::round(scroll[3]->get_value() / 255.0));
-		}
+	Vector<float> components;
+	for (int i = 0; i < current_sliders_count; i++) {
+		components.push_back(scrolls[i]->get_value());
+	}
+	components.push_back(alpha_scroll->get_value());
+
+	color = selected_mode->get_color(components);
 
+	if (mode_option_button->get_selected() == HSV_MODE_INDEX || mode_option_button->get_selected() == OKHSL_MODE_INDEX) {
+		h = scrolls[0]->get_value() / 360.0;
+		s = scrolls[1]->get_value() / 100.0;
+		v = scrolls[2]->get_value() / 100.0;
 		last_color = color;
-	} else {
-		for (int i = 0; i < 4; i++) {
-			color.components[i] = scroll[i]->get_value() / (raw_mode_enabled ? 1.0 : 255.0);
-		}
 	}
 
 	_set_pick_color(color, false);
 	emit_signal(SNAME("color_changed"), color);
 }
 
+void ColorPicker::add_mode(ColorMode *p_mode) {
+	modes.push_back(p_mode);
+	mode_option_button->add_item(RTR(p_mode->get_name()));
+}
+
+HBoxContainer *ColorPicker::get_slider_hbc(int idx) {
+	HBoxContainer *hbc = memnew(HBoxContainer);
+
+	Label *l = memnew(Label());
+	l->set_custom_minimum_size(Size2(get_theme_constant(SNAME("label_width")), 0));
+	l->set_v_size_flags(SIZE_SHRINK_CENTER);
+	hbc->add_child(l);
+
+	HSlider *s = memnew(HSlider);
+	s->set_v_size_flags(SIZE_SHRINK_CENTER);
+	s->set_focus_mode(FOCUS_NONE);
+	hbc->add_child(s);
+
+	SpinBox *v = memnew(SpinBox);
+	s->share(v);
+	hbc->add_child(v);
+	v->get_line_edit()->connect("focus_entered", callable_mp(this, &ColorPicker::_focus_enter));
+	v->get_line_edit()->connect("focus_exited", callable_mp(this, &ColorPicker::_focus_exit));
+
+	s->set_min(0);
+	s->set_page(0);
+	s->set_h_size_flags(SIZE_EXPAND_FILL);
+
+	s->connect("value_changed", callable_mp(this, &ColorPicker::_value_changed));
+	s->connect("draw", callable_mp(this, &ColorPicker::_slider_draw), make_binds(idx));
+
+	if (idx < 4) {
+		scrolls[idx] = s;
+		values[idx] = v;
+		labels[idx] = l;
+	} else {
+		alpha_scroll = s;
+		alpha_value = v;
+		alpha_label = l;
+	}
+	return hbc;
+}
+
+HSlider **ColorPicker::get_sliders() {
+	HSlider **hs = new HSlider *[5];
```

Comment:
I wonder if there's a way to avoid allocating memory on the heap, theoretically if it's just returning 5 pointers, then it shouldn't need to. Probably doesn't matter that much though.